javatests/arcs/core/host/AllocatorTest.kt

```diff
@@ -0,0 +1,222 @@
+    @Retention(AnnotationRetention.RUNTIME)
+    @TargetHost(ReadingHost::class)
+    annotation class RunInReadingHost
+
+    @AutoService(Particle::class)
+    @RunInWritingHost
+    class WritePerson : Particle
+
+    @AutoService(Particle::class)
+    @RunInReadingHost
+    class ReadPerson : Particle
+
+    open class TestingHost : AbstractArcHost() {
+        var started = mutableListOf<PlanPartition>()
+
+        override suspend fun startArc(partition: PlanPartition) {
+            super.startArc(partition)
+            started.add(partition)
+        }
+
+        fun setup() {
+            started.clear()
+        }
+    }
+
+    @AutoService(ArcHost::class)
+    class WritingHost : TestingHost()
+
+    @AutoService(ArcHost::class)
+    class ReadingHost : TestingHost()
+
+    private lateinit var personHandleSpec: HandleSpec
+    private lateinit var readPersonHandleConnectionSpec: HandleConnectionSpec
+    private lateinit var writePersonHandleConnectionSpec: HandleConnectionSpec
+    private lateinit var writePersonParticleSpec: ParticleSpec
+    private lateinit var readPersonParticleSpec: ParticleSpec
+    private lateinit var writeAndReadPersonPlan: Plan
+
+    @Before
+    fun setUp() {
+        runBlocking {
+            personHandleSpec =
+                HandleSpec(
+                    null, "recipePerson", null, mutableSetOf("volatile"),
+                    personSchema
+                )
+
+            writePersonParticleSpec =
+                ParticleSpec("WritePerson", WritePerson::class.java.getCanonicalName()!!)
+            writePersonHandleConnectionSpec =
+                HandleConnectionSpec("person", personHandleSpec, writePersonParticleSpec)
+
+            readPersonParticleSpec =
+                ParticleSpec("ReadPerson", ReadPerson::class.java.getCanonicalName()!!)
+            readPersonHandleConnectionSpec =
+                HandleConnectionSpec("person", personHandleSpec, readPersonParticleSpec)
+
+            writeAndReadPersonPlan = Plan(
+                listOf(writePersonHandleConnectionSpec, readPersonHandleConnectionSpec)
+            )
+
+            ServiceLoaderHostRegistry.availableArcHosts().forEach {
+                when (it) {
+                    is TestingHost -> it.setup()
+                    else -> {
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Tests that the Recipe is properly partitioned so that [ReadingHost] contains only
+     * [ReadPerson] with associated handles and connections, and [WritingHost] contains only
+     * [WritePerson] with associated handles and connections.
+     */
+    @Test
+    fun allocator_computePartitions() = runBlockingTest {
+        val allocator = Allocator(ServiceLoaderHostRegistry)
+        val arcId = allocator.startArcForPlan("readWritePerson", writeAndReadPersonPlan)
+        val planPartitions = allocator.getPartitionsFor(arcId)!!
+        assertThat(planPartitions.size).isEqualTo(2)
```

Comment:
I think you could rewrite this as:

```
assertThat(planPartitions).containsExactly(
    PlanPartition(
        arcId.toString(),
        ReadingHost::class.java.canonicalName,
        listOf(readPersonHandleConnectionSpec)
    ),
    PlanPartition(
        arcId.toString(),
        WritingHost::class.java.canonicalName,
        listOf(writePersonHandleConnectionSpec)
    )
)
```

(not sure if the listOf will pass the equality check correctly, but I believe it will)

Reply:
Nice, I keep forgetting data classes give you equals()/hashCode() for free.